python/lsst/daf/butler/registry/databases/sqlite.py

```diff
@@ -190,6 +180,16 @@ def creator() -> sqlite3.Connection:
 
         engine = sqlalchemy.engine.create_engine(uri, creator=creator)
 
+        def _onSqlite3Begin(connection: sqlalchemy.engine.Connection) -> sqlalchemy.engine.Connection:
+            assert connection.dialect.name == "sqlite"
+            # Replace pysqlite's buggy transaction handling that never BEGINs
+            # with our own that does, and tell SQLite to try to acquire a lock
+            # as soon as we start a transaction (this should lead to more
+            # blocking and fewer deadlocks).
+            if writeable:
```

Comment:
Last night I discovered that we actually do have a use case for read-only transactions for consistency: at butler startup we fetch a lot of information about dataset types and collections, and it's possible for those queries to be made inconsistent by a concurrent writer (and on `main` we don't have a transaction around that at all).  I discovered this while running ci_cpp a lot to try to reproduce an unrelated race condition; I think it's unusual in that it tends to initialize `Butler` instances (for running `PipelineTasks`) while also registering dataset types and collections (when starting to run `PipelineTasks` from a different pipeline), all with SQLite.

So I've pushed a modified version of this commit that just emits `BEGIN` for read-only databases, rather than `BEGIN IMMEDIATE`; that should start a read-only transaction when followed (as it will be) by a `SELECT` query.  The overall behavior should then be:
 - Multiple transactions in read-only databases can occur at the same time, as they only acquire a shared lock on the database file.
 - Any read-write transactions will block until any concurrent read-only or read-write transactions complete.
 - An ongoing read-write transaction will block any read-only transactions from starting.

And I've also wrapped those butler startup fetches in a transaction context.